R-package/src/install.libs.R

```diff
@@ -15,6 +15,39 @@ if (!(R_int_UUID == "0310d4b8-ccb1-4bb8-ba94-d36a55f60262"
   print("Warning: unmatched R_INTERNALS_UUID, may cannot run normally.")
 }
 
+.find_vs_version <- function() {
+  current_working_dir <- getwd()
+  on.exit({
+    setwd(current_working_dir)
+  })
+  vs_versions <- c(
+    "Visual Studio 16 2019"
+    , "Visual Studio 15 2017"
+    , "Visual Studio 14 2015"
+  )
+  working_vs_version <- NULL
+  for (vs_version in vs_versions) {
+    print(sprintf("Trying '%s'", vs_version))
+    build_dir <- tempdir()
+    setwd(build_dir)
+    writeLines(
+      text = "PROJECT(testing)"
+      , con = file.path(build_dir, "CMakeLists.txt")
+    )
+    cmake_cmd <- paste0(
+      "cmake -Wno-dev -G "
+      , shQuote(vs_version)
+      , " ."
```

Comment:
Wasn't 64-bit qualificator `-A x64` lost?

Reply:
whoops yeah you're right! But we don't need it on this line

Reply:
added: https://github.com/microsoft/LightGBM/pull/2965/commits/5e125b92f7f1de59d28052ea920793246a4169ad

testing on my fork: https://github.com/jameslamb/LightGBM/pull/13

Reply:
seems that it succeeded!

AppVeyor: https://ci.appveyor.com/project/jameslamb/lightgbm/builds/31913976
Azure DevOps: https://dev.azure.com/JayLamb20/jameslamb-lightgbm/_build/results?buildId=266&view=logs&jobId=b51fa6ee-4364-59e6-4527-dc4aaafd2db2&j=b51fa6ee-4364-59e6-4527-dc4aaafd2db2&t=85bde6ad-0953-589b-e1e8-633bb2d9d530

Reply:
Unfortunately, those logs cannot help to check whether the build was successful with VS or due to fallback to MinGW. Or am I just searching badly? 🙁

Reply:
oh right! Because the install is happening inside `R CMD check`'s step that is like `checking if 'lightgbm' can be installed...`  and those logs are suppressed.

Let me do a build where I remove the `--skip-install` on `build_r.R` so you can see it.

Reply:
ok try these:

AppVeyor (MinGW): https://ci.appveyor.com/project/jameslamb/lightgbm/builds/31919068/job/66631ptg2ublemex

![image](https://user-images.githubusercontent.com/7608904/78309260-1b874d80-7510-11ea-9410-9f098326e002.png)

Azure DevOps (MSVC from Visual Studio): https://dev.azure.com/JayLamb20/jameslamb-lightgbm/_build/results?buildId=267&view=logs&j=b51fa6ee-4364-59e6-4527-dc4aaafd2db2

![image](https://user-images.githubusercontent.com/7608904/78309372-66a16080-7510-11ea-820d-4fed6f8e1197.png)

^ @StrikerRUS I'm glad you asked! It looks like it is finding the  `gcc.exe` bundled  with chocolatey 😬 . I know that when I was testing before, Visual Studio was found. I will have to test some more :/

Reply:
Also I can say that according to the logs we should remove temp build directory between tries.

Reply:
Yep definitely agree. I'm testing on my fork.

You can see on this commit that MSVC from Visual studio is being used ([build link](https://dev.azure.com/JayLamb20/jameslamb-lightgbm/_build/results?buildId=269&view=logs&j=b51fa6ee-4364-59e6-4527-dc4aaafd2db2))

![image](https://user-images.githubusercontent.com/7608904/78311378-496f9080-7516-11ea-8b21-a110945ea1d2.png)

But I still haven't gotten a successful build. Will `@` you when I do.

Really really glad you asked! I never considered that `gcc` might just be laying around from something else 😬

Reply:
I think I should print the contents of `lightgbm.Rcheck/00install.out` so we can see this in our logs going forward. That's a more permanent fix than removing `--skip-install`

Reply:
ok this did finally succeed but it took *49 minutes*. 

https://dev.azure.com/JayLamb20/jameslamb-lightgbm/_build/results?buildId=273&view=logs&j=b51fa6ee-4364-59e6-4527-dc4aaafd2db2&t=85bde6ad-0953-589b-e1e8-633bb2d9d530

I think I might know what's happening, trying something else.

Reply:
>  I never considered that gcc might just be laying around from something else 😬

I was surprised too!

 >I think I should print the contents of `lightgbm.Rcheck/00install.out` so we can see this in our logs going forward. That's a more permanent fix than removing `--skip-install`

If this is the only way to force displaying logging `print`s, then yes, I agree with you.

> ok this did finally succeed but it took 49 minutes.

Oh, my! 😬

Reply:
Ok I think this is working! But....it takes 26 minutes :/

https://dev.azure.com/JayLamb20/jameslamb-lightgbm/_build/results?buildId=275&view=logs&jobId=b51fa6ee-4364-59e6-4527-dc4aaafd2db2&j=b51fa6ee-4364-59e6-4527-dc4aaafd2db2&t=85bde6ad-0953-589b-e1e8-633bb2d9d530

I think the 20 minutes I referenced in https://github.com/microsoft/LightGBM/pull/2965#issue-397341495 was suffering from silently getting the `gcc.exe` from chocolatey, so it wasn't real :/

I know this is super long. I tried to mitigate the effect on total runtime with two changes in #2936:

* moving `r-package` task first in `.vsts-ci`
* only running `R CMD check --no-manual --ignore-vignettes` to cut out time spent testing documentation (which is already tested on AppVeyor)

I might be able to cut out more time, but either way I don't think that should impact this PR. I think that build link is evidence  that what I'm doing here in this PR does successfully build the R package with MSVC